### PR TITLE
feat: stream csv downloads

### PR DESCRIPTION
**Pull request by muhammad-ammar**
**JIRA:** https://2u-internal.atlassian.net/browse/ENT-8301

**Dependency:** https://github.com/openedx/frontend-app-admin-portal/pull/1167

**Merge checklist:**
- [ ] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `base.in` if needed in production but edx-analytics-data-api doesn't install it
    - `test-master.in` if edx-analytics-data-api pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
- [ ] `make static` has been run to update webpack bundling if any static content was updated
- [ ] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the edx-analytics-data-api requirements installed or if you checked out edx-enterprise-data into the src directory used by edx-analytics-data-api, you can run this command through an edx-analytics-data-api shell.
        - It would be `./manage.py makemigrations` in the shell.
- [ ] [Version](https://github.com/openedx/edx-enterprise-data/blob/master/enterprise/__init__.py) bumped
- [ ] [Changelog](https://github.com/openedx/edx-enterprise-data/blob/master/CHANGELOG.rst) record added
- [ ] Translations updated (see docs/internationalization.rst but also this isn't blocking for merge atm)

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/openedx/edx-enterprise-data/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [Travis](https://travis-ci.org/github/edx/edx-enterprise-data), verify version has been pushed to [PyPI](https://pypi.org/project/edx-enterprise-data/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [edx-analytics-data-api](https://github.com/openedx/edx-analytics-data-api) to upgrade dependencies (including edx-enterprise-data)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-analytics-data-api will look for the latest version in PyPi.
    - Note: the edx-enterprise-data constraint in edx-analytics-data-api **must** also be bumped to the latest version in PyPi.

### CODE REVIEW

File: enterprise_data/api/v1/views.py [hunk 1]
```diff
@@ -140,6 +143,31 @@ def get_queryset(self):
...
+        """
+        queryset = self.filter_queryset(self.get_queryset())
+
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+
+        def data_gen(queryset):
+            paginator = Paginator(queryset, per_page=10000)
```
Comment (iloveagent57) [inline enterprise_data/api/v1/views.py]: small suggestion: make the `10000` value come from settings, so we can tune it up or down without deployment.

File: enterprise_data/api/v1/views.py [hunk 2]
```diff
@@ -140,6 +143,31 @@ def get_queryset(self):
...
+    def list(self, request, *args, **kwargs):
+        """
+        Override the list method to handle streaming CSV download.
+        """
+        queryset = self.filter_queryset(self.get_queryset())
+
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
```
Comment (iloveagent57) [inline enterprise_data/api/v1/views.py]: Do we want this here? Should the block of code that conditionally renders the CSV come first?

File: enterprise_data/api/v1/views.py [hunk 3]
```diff
@@ -140,6 +143,31 @@ def get_queryset(self):
             TieredCache.set_all_tiers(cache_key, enrollments, DEFAULT_LEARNER_CACHE_TIMEOUT)
             return enrollments
 
+    def list(self, request, *args, **kwargs):
+        """
+        Override the list method to handle streaming CSV download.
+        """
```
Comment (iloveagent57) [inline enterprise_data/api/v1/views.py:141]: One idea for rollout: introduce a feature flag, where if the flag is off, this method can probably just return `super().list(...)`. And if it's on, it can do all the new stuff you've introduced.

File: enterprise_data/renderers.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from rest_framework_csv.renderers import CSVStreamingRenderer
+
+
+class EnrollmentsCSVRenderer(CSVStreamingRenderer):
+    header = [
```
Comment (iloveagent57) [inline enterprise_data/renderers.py:17]: You could probably do something like
```python
header = [field.name for field in EnterpriseLearnerEnrollment._meta.get_fields()]
```

File: enterprise_data/api/v1/views.py (hunk 3 above)
Comment (iloveagent57) [reply enterprise_data/api/v1/views.py:141]: Also, here's an example from SO that might give you an idea for how to structure this code a little differently: https://stackoverflow.com/a/65564367 It might help you simplify this a little bit.

File: enterprise_data/api/v1/views.py [hunk 5]
```diff
@@ -140,6 +143,31 @@ def get_queryset(self):
...
+            return self.get_paginated_response(serializer.data)
+
+        def data_gen(queryset):
+            paginator = Paginator(queryset, per_page=10000)
+            for page_number in paginator.page_range:
+                enrollments = paginator.page(page_number)
+                serializer = self.get_serializer(enrollments, many=True)
+                yield serializer.data
+
+        if self.request.query_params.get('data') == 'csv':
```
Comment (iloveagent57) [inline enterprise_data/api/v1/views.py]: Why is this change necessary? Does requesting `...enrollments.csv` no longer work?

Comment (iloveagent57) [reply enterprise_data/api/v1/views.py]: You might be able to do something like this if you want to switch behavior based on whether `.csv` is being requested: https://www.django-rest-framework.org/api-guide/renderers/#varying-behavior-by-media-type

Comment (muhammad-ammar) [reply enterprise_data/api/v1/views.py]: > Why is this change necessary? Does requesting `...enrollments.csv` no longer work?

Unfortunately, yes. I really want to avoid this. I will give it another look.

**Comment by muhammad-ammar** [comment]
@iloveagent57 Bundle of the thanks for the feedback. This work is in progress. I wanted to get your thoughts on the overall approach. I will address all the feedback. 

One question, Things are clear regarding streaming csv. How about chunking the DB reads?

File: enterprise_data/api/v1/views.py (hunk 5 above)
Comment (muhammad-ammar) [reply enterprise_data/api/v1/views.py]: Requests like `enrollments.csv` will now work

File: enterprise_data/api/v1/views.py (hunk 3 above)
Comment (muhammad-ammar) [reply enterprise_data/api/v1/views.py:141]: What do you say if enable/disable old/new functionality based on a query param passed from admin-portal?

Comment (iloveagent57) [reply enterprise_data/api/v1/views.py:141]: Yeah, sure, that's a good idea too.

File: enterprise_data/api/v1/views.py [hunk 6]
```diff
@@ -140,6 +146,30 @@ def get_queryset(self):
...
+
+    def _stream_serialized_data(self):
+        """
+        Stream the serialized data.
+        """
+        queryset = self.filter_queryset(self.get_queryset())
+        serializer = self.get_serializer_class()
+        paginator = Paginator(queryset, per_page=settings.ENROLLMENTS_PAGE_SIZE)
+        for page_number in paginator.page_range:
+            yield from serializer(paginator.page(page_number).object_list, many=True).data
```
Comment (iloveagent57) [inline enterprise_data/api/v1/views.py]: Awesome, nice change to clean this up.

**Review by iloveagent57 — APPROVED** [review]
The code looks good!  Have you been able to test this out on a fake/large data set?

**Comment by muhammad-ammar** [comment]
> The code looks good! Have you been able to test this out on a fake/large data set?

Yes, I am using the dataset for https://portal.edx.org/national-university-singapore/admin/learners for local testing. 

Results with reading chunks of 10000 records from DB

With streaming 
* CSV Size: 76.8 MB
* CSV Rows: 127500
* Average: 33 seconds

Without streaming
* CSV Size: 76.8 MB
* CSV Rows: 127500
* Average:  28 seconds
